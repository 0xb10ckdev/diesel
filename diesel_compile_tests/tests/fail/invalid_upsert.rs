extern crate diesel;

use diesel::prelude::*;

table! {
    users {
        id -> Integer,
        name -> Text,
        hair_color -> Nullable<Text>,
    }
}

fn main() {
    let conn = &mut PgConnection::establish("...").unwrap();

    // this query is valid
    diesel::insert_into(users::table)
        .values(&[(users::id.eq(1), users::name.eq("Sean"))])
        .on_conflict(users::id)
        .filter_target(users::hair_color.eq("black"))
        .do_update()
        .set(users::hair_color.eq("red"))
        .execute(conn)
        .unwrap();

    // This is invalid as filter_target is only valid for
    // do update queries
    diesel::insert_into(users::table)
        .values(&[(users::id.eq(1), users::name.eq("Sean"))])
        .on_conflict(users::id)
        .filter_target(users::hair_color.eq("black"))
        .do_nothing()
        .execute(conn)
        .unwrap();
}
