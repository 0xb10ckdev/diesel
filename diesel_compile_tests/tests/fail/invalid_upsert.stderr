error[E0275]: overflow evaluating the requirement `_: Sized`
    --> tests/fail/invalid_upsert.rs:33:10
     |
33   |         .execute(conn)
     |          ^^^^^^^
     |
     = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`$CRATE`)
     = note: required because of the requirements on the impl of `QueryFragment<_>` for `diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(ColumnInsertValue<columns::id, &diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>, ColumnInsertValue<columns::name, &diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>), users::table>>, users::table, [diesel::query_builder::insert_statement::ValuesClause<(ColumnInsertValue<columns::id, diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>, ColumnInsertValue<columns::name, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>), users::table>; 1], true>, diesel::query_builder::upsert::on_conflict_target_decorations::DecoratedConflictTarget<diesel::query_builder::upsert::on_conflict_target::ConflictTarget<columns::id>, diesel::query_builder::where_clause::WhereClause<diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<columns::hair_color, diesel::expression::bound::Bound<diesel::sql_types::Nullable<diesel::sql_types::Text>, &str>>>>>, diesel::query_builder::upsert::on_conflict_actions::DoNothing>`
     = note: 127 redundant requirements hidden
     = note: required because of the requirements on the impl of `QueryFragment<_>` for `InsertStatement<users::table, diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(ColumnInsertValue<columns::id, &diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>, ColumnInsertValue<columns::name, &diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>), users::table>>, users::table, [diesel::query_builder::insert_statement::ValuesClause<(ColumnInsertValue<columns::id, diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>, ColumnInsertValue<columns::name, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>), users::table>; 1], true>, diesel::query_builder::upsert::on_conflict_target_decorations::DecoratedConflictTarget<diesel::query_builder::upsert::on_conflict_target::ConflictTarget<columns::id>, diesel::query_builder::where_clause::WhereClause<diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<columns::hair_color, diesel::expression::bound::Bound<diesel::sql_types::Nullable<diesel::sql_types::Text>, &str>>>>>, diesel::query_builder::upsert::on_conflict_actions::DoNothing>>`
     = note: required because of the requirements on the impl of `ExecuteDsl<_, _>` for `InsertStatement<users::table, diesel::query_builder::upsert::on_conflict_clause::OnConflictValues<diesel::query_builder::insert_statement::batch_insert::BatchInsert<Vec<diesel::query_builder::insert_statement::ValuesClause<(ColumnInsertValue<columns::id, &diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>, ColumnInsertValue<columns::name, &diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>), users::table>>, users::table, [diesel::query_builder::insert_statement::ValuesClause<(ColumnInsertValue<columns::id, diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>>, ColumnInsertValue<columns::name, diesel::expression::bound::Bound<diesel::sql_types::Text, &str>>), users::table>; 1], true>, diesel::query_builder::upsert::on_conflict_target_decorations::DecoratedConflictTarget<diesel::query_builder::upsert::on_conflict_target::ConflictTarget<columns::id>, diesel::query_builder::where_clause::WhereClause<diesel::expression::grouped::Grouped<diesel::expression::operators::Eq<columns::hair_color, diesel::expression::bound::Bound<diesel::sql_types::Nullable<diesel::sql_types::Text>, &str>>>>>, diesel::query_builder::upsert::on_conflict_actions::DoNothing>>`
note: required by a bound in `diesel::RunQueryDsl::execute`
    --> $DIESEL/src/query_dsl/mod.rs
     |
     |         Self: methods::ExecuteDsl<Conn>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `diesel::RunQueryDsl::execute`
